File: java/yass-base/test/ch/softappeal/yass/util/test/DumperTest.java
package ch.softappeal.yass.util.test;

import ch.softappeal.yass.serialize.test.SerializerTest;
import ch.softappeal.yass.util.Dumper;
import ch.softappeal.yass.util.Nullable;
import ch.softappeal.yass.util.TestUtils;
import org.junit.Ignore;
import org.junit.Test;

import java.io.PrintWriter;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.Instant;
import java.util.Arrays;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.Map;

@Ignore // $$$
public class DumperTest {

    private static final class TestDumper extends Dumper {
        TestDumper(final boolean compact, final boolean referenceables, final Class<?>... concreteValueClasses) {
            super(compact, referenceables, concreteValueClasses);
        }
        @Override protected boolean dumpValueClass(final StringBuilder out, final Class<?> type, final Object object) {
            if (isConcreteValueClass(type) || (Date.class == type)) {
                out.append(object);
                return true;
            }
            return false;
        }
    }

    @Test public void test() {
        TestUtils.compareFile("ch/softappeal/yass/util/test/DumperTest.dump.txt", new TestUtils.Printer() {
            void dump(final Dumper dumper, final StringBuilder s, final @Nullable Object value) {
                s.append(dumper.append(new StringBuilder(256), value)).append('\n');
            }
            void print(final Dumper dumper, final PrintWriter printer, final boolean cycles) {
                final StringBuilder s = new StringBuilder(1024);
                dump(dumper, s, null);
                dump(dumper, s, 'c');
                dump(dumper, s, SerializerTest.createNulls());
                dump(dumper, s, SerializerTest.createValues());
                if (cycles) {
                    dump(dumper, s, SerializerTest.createGraph());
                }
                dump(dumper, s, Arrays.asList("one", "two", "three").toArray());
                final Map<Integer, String> int2string = new LinkedHashMap<>();
                int2string.put(1, "one");
                int2string.put(2, null);
                int2string.put(3, "three");
                dump(dumper, s, int2string);
                printer.append(s);
            }
            @Override public void print(final PrintWriter printer) {
                print(new TestDumper(false, true, BigInteger.class, BigDecimal.class, Instant.class), printer, true);
                print(new TestDumper(false, false), printer, false);
                print(new TestDumper(true, true, BigInteger.class, BigDecimal.class, Instant.class), printer, true);
                print(new TestDumper(true, false), printer, false);
            }
        });
    }

}

File: kotlin/yass/test/ch/softappeal/yass/DumperTest.kt
package ch.softappeal.yass

import ch.softappeal.yass.serialize.createGraph
import ch.softappeal.yass.serialize.createNulls
import ch.softappeal.yass.serialize.createValues
import org.junit.Test
import java.io.PrintWriter
import java.math.BigDecimal
import java.math.BigInteger
import java.time.Instant
import java.util.Date
import java.util.TimeZone
import kotlin.test.assertEquals

private fun dumper(compact: Boolean, graph: Boolean, concreteValueClasses: Set<Class<*>>): Dumper {
    val valueDumper: ValueDumper = { out, value ->
        val type = value.javaClass
        if (
            concreteValueClasses.contains(type) ||
            (Date::class.java === type) ||
            (BigInteger::class.java === type) ||
            (BigDecimal::class.java === type) ||
            (Instant::class.java === type)
        ) out.append(value)
    }
    return if (graph) graphDumper(compact, concreteValueClasses, valueDumper) else treeDumper(compact, valueDumper)
}

class DumperTest {
    @Test
    fun test() {
        val timeZone = TimeZone.getDefault()
        try {
            TimeZone.setDefault(TimeZone.getTimeZone("UTC"))
            compareFile("ch/softappeal/yass/DumperTest.dump.txt") { printer ->
                fun print(dumper: Dumper, printer: PrintWriter, cycles: Boolean) {
                    val s = StringBuilder(1024)
                    fun dump(value: Any?) {
                        dumper(s, value).append('\n')
                    }
                    dump(null)
                    dump('c')
                    dump(createNulls())
                    dump(createValues())
                    if (cycles) dump(createGraph())
                    dump(arrayOf<Any>("one", "two", "three"))
                    dump(mapOf(1 to "one", 2 to null, 3 to "three"))
                    printer.append(s)
                }
                print(
                    dumper(false, true, setOf(BigInteger::class.java, BigDecimal::class.java, Instant::class.java)),
                    printer,
                    true
                )
                print(dumper(false, false, emptySet()), printer, false)
                print(dumper(true, true, setOf(BigInteger::class.java, BigDecimal::class.java, Instant::class.java)), printer, true)
                print(dumper(true, false, emptySet()), printer, false)
            }
        } finally {
            TimeZone.setDefault(timeZone)
        }
    }

    @Test
    fun dump() {
        val dumper = graphDumper(true, setOf(BigInteger::class.java)) { out, value ->
            if (value is BigInteger) out.append(value.toInt() + 1)
        }
        assertEquals("2", dumper(StringBuilder(128), BigInteger.valueOf(1)).toString())
        assertEquals("3", dumper.dump(BigInteger.valueOf(2)).toString())
        assertEquals("4", dumper.toString(BigInteger.valueOf(3)))
    }
}
